src/quaternion_types.jl

```diff
@@ -256,14 +256,31 @@ end
 @inline Base.getindex(spq::SPQuat, i::Int) = convert(Quat, spq)[i]
 @inline Base.Tuple(spq::SPQuat) = Tuple(convert(Quat, spq))
 
+# Optimizations for going between Quat and SPQuat
+@inline (::Type{SPQ})(q::Quat) where {SPQ <: SPQuat} = convert(SPQ, q)
+@inline (::Type{Q})(spq::SPQuat) where {Q <: Quat} = convert(Q, spq)
+
 @inline function Base.convert(::Type{Q}, spq::SPQuat) where Q <: Quat
-    # Both the sign and norm of the Quat is automatically dealt with in its inner constructor
-    return Q(1 - (spq.x*spq.x + spq.y*spq.y + spq.z*spq.z), 2*spq.x, 2*spq.y, 2*spq.z)
+    # Equation (45) in
+    # Terzakis et al., "A Recipe on the Parameterization of Rotation Matrices
+    # for Non-Linear Optimization using Quaternions":
+    alpha2 = spq.x * spq.x + spq.y * spq.y + spq.z * spq.z
+    scale = 2 / (alpha2 + 1)
+    Q((1 - alpha2) / (alpha2 + 1), scale * spq.x, scale * spq.y, scale * spq.z, false)
 end
 
 @inline function Base.convert(::Type{SPQ}, q::Quat) where SPQ <: SPQuat
-    alpha2 = (1 - q.w) / (1 + q.w) # <= 1 since q.w >= 0
-    spq = SPQ(q.x * (alpha2 + 1)*0.5,  q.y * (alpha2 + 1)*0.5, q.z * (alpha2 + 1)*0.5)
+    # Simplification of (46) and (47) in
+    # Terzakis et al., "A Recipe on the Parameterization of Rotation Matrices
+    # for Non-Linear Optimization using Quaternions":
+    # α² = (1 - q.w) / (1 + q.w)
+    # scale = (α² + 1) / 2
+    # scale = ((1 - q.w) / (1 + q.w) + 1) / 2
+    # scale = ((1 - q.w + (1 + q.w)) / (1 + q.w) / 2
+    # scale = (2 / (1 + q.w)) / 2
+    # scale = 1 / (1 + q.w)
+    scale = 1 / (1 + q.w)
```

Comment:
@ryanelandt, I didn't really understand your `copysign` here. Could you check this?

Reply:
Leaving it out was definitely the correct thing.

**Why I put it there:** There was discussion that removing `copysign` from the `Quat` would cause some `SPQuats` to be poorly conditioned. I put a `copysign` there in order to ensure that the `SPQuat` always corresponded to the `Quat` with a positive real part. My point was that perceived issues caused by removing `copysign` can be fixed by writing code slightly more generally.

Reply:
I think `q.w` being close to -1 is still a legitimate concern here. I understand why you had the `copysign` on the `1` in the denominator now. It would result in a discontinuous jump at `q.w == 0` however, which may also be undesirable. On the other hand, that's where the singularity is anyway, so maybe that's alright. @awbsmith, what are your thoughts on this?

Reply:
***Time varying rotations:*** Euler Angles representations of rotations suffer singularities. These singularities are unavoidable without [discrete switching](https://en.wikipedia.org/wiki/Euler_filter). `SPQuats` have a singularity too; but this singularity is easier to resolve. The important thing is that in both cases, the switching/`principle_value` is done at specific times after a finite amount of rotation.

All `SPQuats` associated with `Quats` with non-negative real parts are at least `pi` radians away from the singularity. Therefore, the singularity is 100% avoidable as long as objects rotate less than say `pi/2` during a time-step.

Currently, `SPQuats` are the recommended choice for differentiation/optimization problems. Solving implicit ODEs requires differentiation and optimization; `SPQuats` seem like a logical choice for this task. And they are assuming `copysign` is not there. Adding `copysigns` into constructors requires one to use workarounds when using `SPQuats` in implicit ODEs. Also, `copysign` makes the behavior of `SPQuats` not rational.